### PR TITLE
Add app extension

### PR DESCRIPTION
**Pull request by elland**
Still needs some layout love and a cancel button, but it works.

This also modifies the sorting for the data source and modifies some of the dependencies. If we agree on them, I'll open PRs for the changes in the actual repos.

[Preview](https://cloud.githubusercontent.com/assets/54423/19148124/8b7860f8-8bbb-11e6-8369-78451ab8fbc6.png).

### CODE REVIEW

**Comment by 3lvis** [comment]
Would be awesome to see how it looks :)

**Comment by elland** [comment]
@3lvis it's the same view as the 'add/edit' bookmark view.

File: Vendor/Networking/NetworkActivityIndicator.swift [hunk 1]
```diff
@@ -36,7 +36,9 @@ class NetworkActivityIndicator: NSObject {
             #if os(iOS)
                 let deadline = DispatchTime.now() + Double(Int64(0.5 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)
                 DispatchQueue.main.asyncAfter(deadline: deadline) {
-                    UIApplication.shared.isNetworkActivityIndicatorVisible = (self.activitiesCount > 0)
+                    if UIApplication.responds(to: NSSelectorFromString("shared")) {
```
Comment (3lvis) [inline Vendor/Networking/NetworkActivityIndicator.swift:39]: Why is this needed?

Comment (3lvis) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: People modifying source under `Vendor`. One of the main reasons why dependency managers are a good idea.

Comment (3lvis) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: Don't ever do this, please.

Comment (elland) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: You can't access `shared` from inside an app extension. I thought either this or some `#ifdef`...

Comment (elland) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: Please read the PR description.

Comment (3lvis) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: How would the #ifdef would look like? I remember reading something about people not been able to use something like this in extensions. Is nice to finally know why.

File: Extension/ActionViewController.swift [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+class ActionViewController: UIViewController {
+    lazy var fetcher: Fetcher = {
+        return Fetcher(baseURL: "https://api.pinboard.in/v1")
+    }()
+
+    lazy var toolbar: UIToolbar = {
+        let view = UIToolbar(withAutoLayout: true)
+        view.tintColor = .tint
+
+//        let cancelItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(ActionViewController.cancel))
```
Comment (3lvis) [inline Extension/ActionViewController.swift]: Remove commented code, please.

Comment (elland) [reply Extension/ActionViewController.swift]: Already did.

File: Vendor/Networking/NetworkActivityIndicator.swift (hunk 1 above)
Comment (elland) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: Like the ones you use on Sweet\* stuff?

Comment (elland) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: But I'd have too check what `KEY` I'd have to check for.

Comment (3lvis) [reply Vendor/Networking/NetworkActivityIndicator.swift:39]: Yeah, I was mainly asking because I'm not even sure if it's possible. What you did here might be the best solution.

**Comment by elland** [comment]
Updated description to add preview.

File: Extension/ActionViewController.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        self.addConstraints()
+        
+        // For example, look for an image and place it into an image view.
+        // Replace this with something appropriate for the type[s] your extension supports.
+        var urlFound = false
+        for item in self.extensionContext!.inputItems as! [NSExtensionItem] {
+            for provider in item.attachments! as! [NSItemProvider] {
+                if provider.hasItemConformingToTypeIdentifier(kUTTypeURL as String) {
+                    // This is an image. We'll load it, then place it in our image view.
+                    provider.loadItem(forTypeIdentifier: kUTTypeURL as String, options: nil) { (url, error) in
```
Comment (3lvis) [inline Extension/ActionViewController.swift]: Please remove parenthesis around `(url, error)`.

**Comment by 3lvis** [comment]
Reviewing just got more difficult. Could you undo the previous commit and lint in a separate PR?

File: Extension/ActionViewController.swift (hunk 3 above)
Comment (elland) [reply Extension/ActionViewController.swift]: Done in separate PR

**Comment by elland** [comment]
🙆

**Comment by elland** [comment]
¿puedo barco?

File: Extension/ActionViewController.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                            self.loadData(with: url)
+                        }
+                    }
+                    
+                    urlFound = true
+                    break
+                }
+            }
+            
+            if (urlFound) {
```
Comment (3lvis) [inline Extension/ActionViewController.swift]: Remove the parenthesis from here as well, please.

File: Extension/ActionViewController.swift [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+    }
+
+    func cancel() {
+        self.extensionContext!.completeRequest(returningItems: nil)
+    }
+
+    func done() {
+        // Create new bookmark for Pinboard
```
Comment (3lvis) [inline Extension/ActionViewController.swift]: I would drop this comment and make the method name more understandable. 

``` swift
func createNewBookmark() {
```

File: Extension/ActionViewController.swift (hunk 4 above)
Comment (elland) [reply Extension/ActionViewController.swift]: Different PR doing that

File: Extension/ActionViewController.swift [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        guard let path = self.pathLabel.text else { return }
+        let parameters: [String: String] = ["description": self.descriptionTextField.text ?? "", "tags": self.tagsTextField.text ?? ""]
+
+        self.fetcher.addBookmark(with: path, parameters: parameters) { error in
+            if let error = error {
+                let alert = UIAlertController.errorAlert(error)
+                self.present(alert, animated: true)
+            } else {
+                self.extensionContext!.completeRequest(returningItems: nil)
```
Comment (3lvis) [inline Extension/ActionViewController.swift:160]: Wow, where does the `extensionContext` comes from? UIViewController?

Comment (elland) [reply Extension/ActionViewController.swift:160]: That's from the application extension stuff.

**Comment by 3lvis** [comment]
Looks good! Please merge it 💐

**Comment by elland** [comment]
Updated.

**Comment by 3lvis** [comment]
Since both were already approved, I think is fine you just merge now. Or did you do additional changes?

**Comment by elland** [comment]
No, just wanted another pair of eyes to check I didn't miss anything.
